README.md
# Under re-generating...

To see change detail, refer to [CHANGELOG.md](./CHANGELOG.md)

Change Category:

+ **A**: edit coding style (follow PEP 8 and tensorflow/research/deeplab)

+ **B**: update code/function for using 

+ **C**: update repository architecture

+ **D**: DEBUG

## Getting-Start
### e.g. Carvana data
+ Step1. Download Carvana dataset from kaggle
+ Step2. transform format of Carvana mask from `gif` to `jpg`
```shell
python3 transform_images.py --dir ./Carvana/train_masks
```
+ Step3. build tfrecord of Carvana dataset
```shell
python3 build_Carvana_to_tfrecord.py --train_image_folder ./Carvana/train --train_label_folder ./Carvana/train_masks_jpg --build_datadir ./Carvana
```
```shell
python3 tfrecord2image_carvana_format.py --tfrecord_path datasets/Carvana/Carvana_train.tfrecord
```
+ Step4. setting parameter to train the model

Recommend models directory naming: Target_Arch_lossName_datasetName_InputDimenstion_InputWidth_InputHeight_version

e.g. CarSeg_UNet_CE_Carvana_2D_512_512_v1
```shell
# this is for testing, not the best settings
python3 train.py --train_logdir models/CarSeg_UNet_CE_Carvana_2D_512_512_v1 --phase train --dataset_dir ./datasets/Carvana --dataset Carvana_train.tfrecord --train_batch_size 2 --training_number_of_steps 30 --summary_steps 100 --save_steps 100 --checkpoint_steps 2
```

### another dataset
+ Step1. Download your dataset, and figure out data type and what data you want to store in tfrecord
+ Step2. create a new build function `build_*New Dataset Name*_to_tfrecord.py` in `datasets/`
+ Step3. follow build function of other dataset to generate tfreocrd of new dataset.


# Unet 
* Tensorflow implement of U-Net: Convolutional Networks for Biomedical Image Segmentation..[[Paper]](https://arxiv.org/abs/1505.04597)
* Borrowed code and ideas from zhixuhao's unet: https://github.com/zhixuhao/unet.

## Install Required Packages
First ensure that you have installed the following required packages:
* TensorFlow1.4.0 ([instructions](https://www.tensorflow.org/install/)). Maybe other version is ok.
* Opencv ([instructions](https://github.com/opencv/opencv)). Here is opencv-2.4.9.

## Datasets
* In this implementation of the Unet, we use Carvana Image Masking Challenge data.[[download]](https://www.kaggle.com/c/carvana-image-masking-challenge/data) We download train.zip and train_masks.zip. You can put all the datasets in datasets folder.
* Run **scripts/transform_images.py** to transform all the image to gray JPEG image.
* Run **scripts/build_tfrecords.py** to generate training data, data format is tfrecords.

## Training and Testing Model
* Run the following script to train the model, in the process of training, will save the training images every 500 steps. See the **model/unet.py** for details.
```shell
sh train.sh
```
You can change the arguments in train.sh depend on your machine config.
* Run the following script to test the trained model. The test.sh will transform the datasets.
```shell
sh test.sh
```
The script will load the trained StarGAN model to generate the transformed images. You could change the arguments in test.sh depend on your machine config.

## Downloading data/trained model
* Pretrained model: [[download]](https://drive.google.com/open?id=14_8ZthgcpIXdEQEzIENueXv7dGVzHvjK). The model-8500 is better.

CHANGELOG.md
# CHANGELOG


## Model
### 05-26-2020
#### [CarSeg_UNet_CE_2D_512_512_v1](./logs/CarSeg_UNet_CE_Carvana_2D_512_512_v1.log)

Dataset: Carvana

## code logs
Change Category:
+ **A**: edit coding style (follow PEP 8 and tensorflow/research/deeplab)
+ **B**: update code/function for using 
+ **C**: update repository architecture
+ **D**: DEBUG

### May 28, 2020
+ B-7: 修改`datasets/build_BraTS_2019_to_tfrecord.py`變數名稱
+ B-8: `model/read_tfrecords.py`的`image_label`改為不除以255
+ D-2: 在`model/unet.py`加入檢查model output的min/max值
+ B-9: 修改`model/utils.py`的儲存訓練快照格式
+ C-4: 新增`tfrecord2image_brats_2019_format.py`
+ ~D-3: 補除以1.0，因tf.cast(data, tf.float32)沒有真的將data轉為float型態~
+ D-3: 呼叫`model/unet.py`中的`self.output.eval()`出錯

Future work:
+ B-10: 加入validation資料
+ C-5: 解構`model/read_tfrecord.py`，使可適應新增資料集
+ D-4: input mask有負值正常嗎？

### May 27, 2020
+ D-1: 發現transform_image.py並沒有將gif正確轉至jpg，label不是[0, 255]，而是漸變(不論使用pyplot, pillow, cv2儲存再讀取皆會出錯)，原因是因為jpg是有損壓縮，儲存後再載入會有差異，將label格式轉為png
+ C-2: 把`train_results/`資料夾移至`models/*Train Model*/`
+ B-4: 改變`model/utils.py`儲存訓練快照function(cv2 -> plt)
+ B-5: 將`model/models.py`最後一層輸出加入sigmoid activation function
+ B-6: 將`scripts/transform_images.py`中，label值255改為1

Future work:
+ B-3(C-3): 把`train_results/`改為用tensorboard顯示

### May 26, 2020
+ A-1: 模型資料夾名稱改為 目的_架構_lossName_datasetName_InputDimenstion_InputWidth_InputHeight_(InputDepth_)version
+ B-1: 把每次參數設定加進logging info
+ C-1: 把`Unet.log`獨立至`log/`


Future work:
+ B-2: 修改input image比例, 並對應修改input mask, output mask的shape

train.py
'''An implementation of CycleGan using TensorFlow (work in progress).'''
import os
import logging

import tensorflow as tf
import numpy as np
import cv2
import scipy.misc # save image

from model import unet

FLAGS = tf.app.flags.FLAGS

tf.app.flags.DEFINE_string('train_logdir', 
                           'models/Target_Arch_lossName_datasetName_InputDimenstion_InputWidth_InputHeight_version', 
                           'Where the checkpoint and logs are stored.')

tf.app.flags.DEFINE_string('phase', 'train', 
                           'model phase: train/test.')

tf.app.flags.DEFINE_string('dataset_dir', './datasets', 
                           'dataset path for training.')

tf.app.flags.DEFINE_string('dataset', 'Carvana_train.tfrecord',
                           'Name of the segmentation dataset.')

tf.app.flags.DEFINE_string('test_dataset', './datasets/test', 
                           'dataset path for testing one image pair.')

tf.app.flags.DEFINE_integer('train_batch_size', 64, 
                            'The number of images in each batch during training.')

tf.app.flags.DEFINE_integer('training_number_of_steps', 100000, 
                            'The number of steps used for training')

tf.app.flags.DEFINE_integer('summary_steps', 100, 
                            'summary period.')

tf.app.flags.DEFINE_integer('checkpoint_steps', 1000, 
                            'checkpoint period.')

tf.app.flags.DEFINE_integer('save_steps', 500, 
                            'checkpoint period.')

tf.app.flags.DEFINE_string('checkpoint', None, 
                           'checkpoint name for restoring.')

tf.app.flags.DEFINE_string('tf_initial_checkpoint_dir', None,
                           'pretrained model directory.')

def main(_):
    tf.logging.set_verbosity(tf.logging.INFO)
    # gpu config.
    config = tf.ConfigProto()
    #config.gpu_options.per_process_gpu_memory_fraction = 0.5
    # config.gpu_options.allow_growth = True

    if FLAGS.phase == 'train':
        with tf.Session(config=config) as sess: 
        # when use queue to load data, not use with to define sess
            train_model = unet.UNet(sess, FLAGS, 
                                    is_training=True)
            train_model.train(batch_size=FLAGS.train_batch_size, 
                              training_number_of_steps=FLAGS.training_number_of_steps, 
                              summary_steps=FLAGS.summary_steps, 
                              checkpoint_steps=FLAGS.checkpoint_steps, 
                              save_steps=FLAGS.save_steps,
                              dataset_dir=FLAGS.dataset_dir,
                              dataset=FLAGS.dataset,
                              tf_initial_checkpoint_dir=FLAGS.tf_initial_checkpoint_dir)
    else:
        with tf.Session(config=config) as sess:
            # test on a image pair.
            test_model = unet.UNet(sess, FLAGS)
            test_model.load(FLAGS.checkpoint)
            image, output_masks = test_model.test()
            # return numpy ndarray.
            
            # save two images.
            print('\n\n\n\n???\n\n\n\n')
            filename_A = 'input.png'
            filename_B = 'output_masks.png'
            
            cv2.imwrite(filename_A, np.uint8(image[0].clip(0., 1.) * 255.))
            cv2.imwrite(filename_B, np.uint8(output_masks[0].clip(0., 1.) * 255.))

            # Utilize cv2.imwrite() to save images.
            print('Saved files: {}, {}'.format(filename_A, filename_B))

if __name__ == '__main__':
    if not os.path.exists('./models'):
        os.makedirs('./models')
    if not os.path.exists('./logs'):
        os.makedirs('./logs')
    logging.basicConfig(
        format='%(asctime)s [%(levelname)s] %(message)s',
        filename='./logs/' + FLAGS.train_logdir.split('/')[-1] + '.log',
        level=logging.DEBUG,
        filemode='w'
    )
    logging.info(
        (
            '--train_logdir={0} \\\n' +
            '--phase={1} \\\n' +
            '--dataset_dir={2} \\\n' +
            '--dataset={3} \\\n' +
            '--test_dataset={4} \\\n' +
            '--train_batch_size={5} \\\n' +
            '--training_number_of_steps={6} \\\n' +
            '--summary_steps={7} \\\n' +
            '--checkpoint_steps={8} \\\n' +
            '--save_steps={9} \\\n' +
            '--checkpoint={10} \\\n' +
            '--tf_initial_checkpoint_dir={10} \\\n'
        ).format(
            FLAGS.train_logdir,
            FLAGS.phase,
            FLAGS.dataset_dir,
            FLAGS.dataset,
            FLAGS.test_dataset,
            FLAGS.train_batch_size,
            FLAGS.training_number_of_steps,
            FLAGS.summary_steps,
            FLAGS.checkpoint_steps,
            FLAGS.save_steps,
            FLAGS.checkpoint,
            FLAGS.tf_initial_checkpoint_dir
        )
    )
    #logging.getLogger().addHandler(logging.StreamHandler())
    tf.app.run()
